doc/design/fluid_compiler.md

```diff
@@ -0,0 +1,110 @@
+# PaddlePaddle Fluid: Towards a Compiled Programming Language
+
+As described in [fluid.md](fluid.md), when a Fluid application program
+runs, it generates a `ProgramDesc` protobuf message as an intermediate
+representation of itself.  The C++ class `Executor` can run this
+protobuf message as an interpreter.  This article describes the Fluid
+compiler.
+
+![](fluid-compiler.png)
+
+## ProgramDesc
+
+Before we go deeper into the idea of compiled language, let us take a
+look at a simple example Fluid application.
+
+```python
+import "fluid"
+
+func paddlepaddle() {
+  X = fluid.read(...)
+  W = fluid.Tensor(...)
+  Y = fluid.mult(X, W)
+}
+```
+
+This program consists of a [block](block.md) of three operators --
+`read`, `assign`, and `mult`.  Its `ProgramDesc` message looks like
+the following
+
+```protobuf
+message ProgramDesc {
```

Comment:
From my understanding the ProgramDesc is a intermediate representation (IR), currently we have Python as frontend that generates the IR, and this PR discusses a cpp code backend.

I think having Python as a frontend is a huge pain. In my opinion the benefit of Python in the machine learning field is:

1. libraries such as numpy
1. Python native program control primitives such as `for`, `if` to control the training process, the researchers are familiar with them.

In our case we are benefit from neither of these two points:

1. the transpiled code can not use numpy.
1. the transpiled code can not use Python native program control primitives.

And we are trapped in the Python grammar.

I think a better way is to invent our own language.

Reply:
I agree and I believe that a new language is the future.